plumpy/base/state_machine.py

```diff
@@ -125,11 +125,11 @@ class State:
     def is_terminal(cls) -> bool:
         return not cls.ALLOWED
 
-    def __init__(self, state_machine: 'StateMachine', *args: Any, **kwargs: Any):  # pylint: disable=unused-argument
+    def __init__(self, state_machine: Optional['StateMachine'], *args: Any, **kwargs: Any):  # pylint: disable=unused-argument
         """
         :param state_machine: The process this state belongs to
         """
-        self.state_machine = state_machine
+        self.state_machine: Optional[StateMachine] = state_machine
```

Comment:
I feel like a `ValueError` might be useful to be riased if `state_machine` is `None` for non terminal states.  What do you think?

If it mattered more I would almost change the class hierarchy to have a `TerminalState` which doesn't even take it as an argument and is hardcoded to return `False` on `is_terminal` but I don't think it's worth it in this case but the exception might be useful.  Just because I don't know how hardcoded the assumption is that the `state_machine` returns a non-None.